File: BestTimeBuySellStock121.java
/**
 * URL of problem:
 * https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
 */

public class BestTimeBuySellStock121 {
    public static void main(String[] args) {
        Solution121 sol = new Solution121();
        int[] prices = {7, 1, 5, 3, 6 ,4};
        System.out.println(sol.maxProfit(prices));
    }
}

class Solution121 {
    public int maxProfit(int[] prices) {
        if(prices.length == 0) {
            return 0;
        }
        int arrSize = prices.length;
        int[] bestProfit = new int[arrSize];
        bestProfit[0] = -1*prices[0];
        int dayShouldBuy = 0;
        for(int i = 1; i < arrSize; i++) {
            int buyToday = -1*prices[i];
            int sellToday = -1*prices[dayShouldBuy] + prices[i];
            int doNothing = bestProfit[i-1];
            if(buyToday >= sellToday && buyToday >= doNothing) {
                dayShouldBuy = i;
                bestProfit[i] = buyToday;
            }
            else {
                bestProfit[i] = Math.max(buyToday, Math.max(sellToday, doNothing));
            }

            if(sellToday > buyToday && sellToday > doNothing && sellToday < 0) {
                bestProfit[i] = buyToday;
                dayShouldBuy = i;
            }
        }
        return Math.max(0, bestProfit[arrSize-1]);
    }
}

File: best_time_buy_sell_stock_121.py
"""
URL of problem:
https://leetcode.com/problems/best-time-to-buy-and-sell-stock/
"""


class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        max_profit = 0
        for i, purchase_price in enumerate(prices[:-1]):
            j = i+1
            for sell_price in prices[i+1:]:
                if sell_price - purchase_price > max_profit:
                    max_profit = sell_price - purchase_price

        return max_profit


def main():
    print(Solution().maxProfit([7, 1, 5, 3, 6, 4]))


if __name__ == '__main__':
    main()
